src/core_plugins/kibana/public/visualize/editor/editor.js

```diff
@@ -66,16 +66,27 @@ uiModules
     location: 'Visualization Editor'
   });
 
+  // Retrieve the resolved SavedVis instance.
   const savedVis = $route.current.locals.savedVis;
 
+  // Instance of vis.js.
   const vis = savedVis.vis;
+
+  // Clone the _vis instance.
   const editableVis = vis.createEditableVis();
+
+  // NOTE: Why is this instance method being overwritten?
   vis.requesting = function () {
     const requesting = editableVis.requesting;
+    // Invoking requesting() calls onRequest on each agg's type param.
+    // NOTE: What is the structure of an agg_config, and the role of type
+    // and type.params?
     requesting.call(vis);
     requesting.call(editableVis);
   };
 
+  // Boolean by default, but then something else later
+  // NOTE: What is the role of searchSource? Why does it default to a boolean?
```

Comment:
*this is kind of a scattershot description, I'll probably come back and edit it to be more coherent later*

`SearchSource`s are essentially promise-based streams of search results that can inherit from other search sources.

Because filters/queries in Kibana have different levels of persistence and come from different places, it is important to keep track of where filters come from for when they are saved back to the `savedObject` store in the Kibana index. To do this, we create trees of `searchSource` objects that can have associated query parameters (index, query, filter, etc) which can also inherit from other `searchSource` objects.

At query time, all of the `searchSource` objects that have subscribers are "flattened", at which point the query params from the `searchSource` are collected while traversing up the inheritance chain. At each link in the chain a decision about how to merge the query params is made until a single set of query parameters is created for each active `searchSource` (a `searchSource` with subscribers). 

That set of query parameters is then sent to elasticsearch. This is how the filter hierarchy works in Kibana.

---
Visualize, starting from a new search:

  1. the `savedVis.searchSource` is set as the `appSearchSource`(currently managed [here](http://github.com/spalger/kibana/blob/01cbb4765a3bab539b2d3abeb60b0ac8af304618/src/ui/public/courier/data_source/_root_search_source.js#L47))
  1. The `savedVis.searchSource` would normally inherit from the `appSearchSource`, but now it is upgraded to inherit from the `rootSearchSource`
  1. Any interaction with the visualization will still apply filters to the `appSearchSource`, so they will be stored directly on the `savedVis.searchSource`
  1. Any interaction with the time filter will be written to the `rootSearchSource`, so those filters will not be saved by the `savedVis`
  1. When the `savedVis` is saved to elasticsearch, it takes with it all the filters that are defined on it directly, but none of the ones that it inherits from other places.

---
Visualize, starting from an existing search:

 1. The `savedVis` loads the `savedSearch` that it is built on top of
 1. The `savedVis.searchSource` is set to inherit from the `saveSearch.searchSource` and set as the `appSearchSource`
 1. the `savedSearch.searchSource`, is set to inherit from the `rootSearchSource`
 1. when the `savedVis` is written to elasticsearch it will be flattened and only include the filters created in the visualize application and will reconnect the filters from the `savedSearch` at runtime to prevent losing the relationship

---
Dashboard search sources:

  1. each panel in a dashboard has a search source
  1. the `savedDashboard` also has a searchsource, and it is set as the `appSearchSource`
  1. each panel's search source inherits from the `appSearchSource`, meaning that they inherit from the dashboard search source.
  1. when a filter is added to the search box, or via a visualization, it is written to the `appSearchSource`